docs/proposals/service-catalog.md

```diff
@@ -0,0 +1,143 @@
+
+## Dynamic provisioning of cluster resources
+
+A user creates a "template" (e.g. if something similar to [OpenShift Templates](https://docs.openshift.org/latest/dev_guide/templates.html) exists) that makes it easy to create everything needed to spin up a new PostgreSQL database (customizable username/password, Service, Deployment, etc.). The user wants to share only this template in the service catalog so others can find it and use it, while keeping other templates in the namespace private.
+
+
+## Custom dynamic provisioning
+
+The IT department manages an off-cluster database. Each developer wanting to access the database is required to use a unique username and password to access the database. Additionally, each developer accesses a unique tablespace that no other team members can access. The IT department used to create database accounts and tablespaces by hand in response to individual developer requests.
+
+Moving forward, IT wants to automate the process. They create a web service that implements the "service broker" API (such as the one from [Cloud Foundry](http://docs.cloudfoundry.org/services/api.html#api-overview)). They add an entry to the service catalog for their database service, pointing at their web service. When a developer consumes the entry from the catalog, IT's web service is contacted, resulting in a new username, password, and tablespace.
+
+# Service configuration data
+
+The only data currently associated with a service is a list of zero or more endpoints. Many services also have configuration data associated with them. These include things such as logins, passwords, and additional connection parameters. To tie configuration data to a service, we propose adding information to the Service type to be able to express a relationship between the Service and its relevant configuration data. This could be accomplished by adding new field(s) or by using annotations. Potential configuration data target reference types include Secrets and ConfigData.
+
+Just creating the reference association by itself does not accomplish much; to be useful, that data needs to be available to the processes executing in containers. More on this process is below in the sections on service claims and linking.
+
+# Service catalogs
+
+A service catalog is a listing of "services". Examples might include an external database that is deployed outside of the cluster, an actual Kubernetes Service, a template of cluster primitives to be created, and custom provisioners that can perform tasks such as creating users in external systems.
```

Comment:
Hrm - I don't want to design too close to assuming that only existing
services can be in the catalog.  One key point for any extensible model is
that the mechanism to identify the target (in a catalog) is generally the
thing you pass to a claim - make the wrong assumption there, and I think
the catalog becomes less valuable.

There are two types of catalog:

1. Retrieving services that you have access to directly (catalog is the
existing service endpoint)
2. Retrieving services that you don't have access to, or aren't really
services

The former is clearly an object reference - user A wishes to claim Service
B in Namespace C.

In the latter, object references could still work - the catalog endpoint
returns a reference to a service that exists.  But this is why I think
allowing references to non-existent things becomes an important design
point - if you only ever reference services that exist in the latter model,
you don't really need a catalog of services, you are just looking for an
alternative view of services that bypasses / sidesteps access control.

Any "sharing" case for multi-tenant services requires what I consider
provisioning - the creation of a new secret, account, or client certificate
as real resources for the user.  The controller(s) that satisfy that are
provisioner controllers.  Sharing also raises the need to consider whether
we should expose the origin service itself - it may be inappropriate for a
consumer to know the mechanical details of how that origin service is
satisfied (labels, selectors, etc).  To satisfy that model the catalog
needs to slightly abstract the origin service - present it to the consumer
in a way that is immediately useful.  That abstraction should not be so
bound to services that it is impossible to entertain the idea that there is
no backing service (account on a third party system).


On Nov 23, 2015, at 4:48 PM, Gabe Monroy <notifications@github.com> wrote:

In docs/proposals/service-catalog.md
<https://github.com/kubernetes/kubernetes/pull/17543#discussion_r45665643>:

> +
> +## Custom dynamic provisioning
> +
> +The IT department manages an off-cluster database. Each developer wanting to access the database is required to use a unique username and password to access the database. Additionally, each developer accesses a unique tablespace that no other team members can access. The IT department used to create database accounts and tablespaces by hand in response to individual developer requests.
> +
> +Moving forward, IT wants to automate the process. They create a web service that implements the "service broker" API (such as the one from [Cloud Foundry](http://docs.cloudfoundry.org/services/api.html#api-overview)). They add an entry to the service catalog for their database service, pointing at their web service. When a developer consumes the entry from the catalog, IT's web service is contacted, resulting in a new username, password, and tablespace.
> +
> +# Service configuration data
> +
> +The only data currently associated with a service is a list of zero or more endpoints. Many services also have configuration data associated with them. These include things such as logins, passwords, and additional connection parameters. To tie configuration data to a service, we propose adding information to the Service type to be able to express a relationship between the Service and its relevant configuration data. This could be accomplished by adding new field(s) or by using annotations. Potential configuration data target reference types include Secrets and ConfigData.
> +
> +Just creating the reference association by itself does not accomplish much; to be useful, that data needs to be available to the processes executing in containers. More on this process is below in the sections on service claims and linking.
> +
> +# Service catalogs
> +
> +A service catalog is a listing of "services". Examples might include an external database that is deployed outside of the cluster, an actual Kubernetes Service, a template of cluster primitives to be created, and custom provisioners that can perform tasks such as creating users in external systems.

I prefer we limit scope of the catalog to services and exclude things that
don't exist yet like templates/provisioners.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17543/files#r45665643>.